### PR TITLE
Add torch.__config__.show(), reporting detailed version of all libraries.

### LINKED ISSUE #111: numpy.__config__.show() for torch
Would be useful to have stats like BLAS, CUDA archs, compile options etc at runtime, similar to `numpy.__config__.show()`

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18579 Add torch.__config__.show(), reporting detailed version of all libraries.**

Fixes #111.

Example output:

```
PyTorch built with:
  - GCC 4.8
  - Intel(R) Math Kernel Library Version 2019.0.3 Product Build 20190125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201107 (a.k.a. OpenMP 3.1)
  - NNPACK is enabled
  - CUDA Runtime 9.2
  - NVCC architecture flags: -gencode;arch=compute_52,code=sm_52
  - CudNN 7.10.2  (built against CUDA 9.1)
  - Magma 2.5.0
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS=  -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-unused-but-set-variable -Wno-maybe-uninitialized, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=True, USE_NNPACK=True, USE_OPENMP=ON,
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gh-metadata: pytorch pytorch 18579 gh/ezyang/48/head

Differential Revision: [D14778507](https://our.internmc.facebook.com/intern/diff/D14778507)

### CODE REVIEW

**Comment by soumith** [comment]
this fixes https://github.com/pytorch/pytorch/issues/111

You might want to rename the methods slightly to match numpy, if you think it makes sense

**Comment by ezyang** [comment]
Makes sense to me!

File: torch/csrc/Module.cpp [hunk 1]
```diff
@@ -299,6 +299,13 @@ PyObject *THPModule_hasDistributed(PyObject *_unused)
 #endif
 }
 
+static PyObject *THPModule_detailedVersion(PyObject *module)
```
Comment (dzhulgakov) [inline torch/csrc/Module.cpp]: nit: shall we start using pybind universally?

**Review by dzhulgakov — APPROVED** [review]
Would be nice to unify it with Caffe2's build options here: caffe2/core/macros.h.in caffe2/core/common.cc

Specifically, if you follow new format, I'd add the following compilation flags:
- BUILD_TYPE
- CXX_FLAGS
- BLAS (as we will likely replace the ugly THBlas soon)
- USE_NNPACK
- USE_FBGEMM (as we're adding quantization support now) - and include its version

File: torch/csrc/Module.cpp (hunk 1 above)
Comment (apaszke) [reply torch/csrc/Module.cpp]: For things like this one sure, but I don't trust pybind enough to feel ok with using it for our performance-critical eager functions.

**Comment by ezyang** [comment]
> Would be nice to unify it with Caffe2's build options here: caffe2/core/macros.h.in caffe2/core/common.cc

Thanks, was not aware of this.

**Comment by ezyang** [comment]
> USE_FBGEMM (as we're adding quantization support now) - and include its version

fbgemm does not appear to have any useful versions. Determined by looking at `Fbgemm.h` and searching for `configure_file` in the cmake

**Comment by ppwwyyxx** [comment]
For cuda/cudnn, etc, could build-time version be different from run-time version in its minor version number or patch number?
If yes, it would be great to have access to both build-time version and run-time version available for debugging (perhaps with `cudnnGetVersion` & `cudaRuntimeGetVersion`)

**Comment by ezyang** [comment]
@ppwwyyxx Makes sense to me!

**Comment by ezyang** [comment]
Now also with nvcc arch information

**Review by dzhulgakov — COMMENTED** [review]
So, do you want to also include the following:
BUILD_TYPE
CXX_FLAGS
BLAS (as we will likely replace the ugly THBlas soon)
USE_NNPACK
USE_FBGEMM (as we're adding quantization support now) - and include its version

Even if FBGEMM doesn't have exported version, just the fact whether we build with it or not, same for NNPACK.

Even for MKL and others - I'd suggest to explicitly adding a line "MKL: none" to the output

**Comment by ezyang** [comment]
@dzhulgakov I'm assuming part of your comment is an accidental copy paste from the earlier comment.

Right now, I print all of the existing Caffe2 settings as part of the `Build settings` key, filtering out keys which are set to an empty string. Do you want each of these to be given their own line in the output? Also, are you asking to stop filtering out empty string keys? (I feel they add unnecessary noise to the output)

**Review by dzhulgakov — APPROVED** [review]
Apologies, I was looking at the old code version (GH UI is confusing)

**Comment by vadimkantorov** [comment]
would this be useful for `collect_env` script?

**Comment by ppwwyyxx** [comment]
@ezyang 
>   - CudNN 7.10.2  (built against CUDA 9.1)

There is no cudnn 7.10.2.

Cudnn version numbers work a bit different from cuda. 7102 means 7.1.2.

**Comment by ezyang** [comment]
@ppwwyyxx Oops! Thanks for the catch.

**Pull request closed without merging**